### PR TITLE
Devsh

### PR DESCRIPTION
**Pull request by Ivanshelemeh**
fixed part of customfield module

### CODE REVIEW

File: src/main/java/com/atlassian/shele/shelePlugin/customfield/SecondCustomField.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        super(customFieldValuePersister, genericConfigManager);
+        this.manager = manager;
+    }
+
+    @Nonnull
+    @Override
+    public Map<String, Object> getVelocityParameters(Issue issue, CustomField field,
+                                                     FieldLayoutItem fieldLayoutItem) {
+        Map<String, Object> params = new HashMap<>();
+        String report = issue.getReporter().getKey();
```
Comment (pavel-fando) [inline src/main/java/com/atlassian/shele/shelePlugin/customfield/SecondCustomField.java]: use name like `repoterKey` or `reporter`

File: src/main/java/com/atlassian/shele/shelePlugin/customfield/SecondCustomField.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        params.put("report", report);
+        return params;
+    }
+
+    @Override
+    public String getValueFromIssue(CustomField field, Issue issue) {
+        int countAssignee = 0;
```
Comment (pavel-fando) [inline src/main/java/com/atlassian/shele/shelePlugin/customfield/SecondCustomField.java]: you could remove this variable and use `lengthAssignee`

File: src/main/java/com/atlassian/shele/shelePlugin/customfield/SecondCustomField.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        List<ChangeItemBean> itemBeans = ComponentAccessor.getChangeHistoryManager()
+                .getChangeItemsForField(issue, CUST_FIELD);
+        int lengthAssignee = itemBeans.size();
+        countAssignee = lengthAssignee;
```
Comment (pavel-fando) [inline src/main/java/com/atlassian/shele/shelePlugin/customfield/SecondCustomField.java]: this line is not neccessary, you can use `lengthAssignee` directly

File: src/main/java/com/atlassian/shele/shelePlugin/utilit/Utilities.java [hunk 4]
```diff
@@ -11,4 +11,5 @@
     public final static String EVENT = "event";
     public final static String ENT = "entity";
     public final static ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+    public final static String CUST_FIELD="assignee";
```
Comment (pavel-fando) [inline src/main/java/com/atlassian/shele/shelePlugin/utilit/Utilities.java]: format line

File: src/main/resources/atlassian-plugin.xml [hunk 5]
```diff
@@ -62,7 +62,12 @@
         <resource type="velocity" name="edit" location="/template/edit.vm"/>
         <resource type="velocity" name="view" location="/template/custom.vm"/>
     </customfield-type>
-
+    <customfield-type key="second-custom-field" name="Second custom field" i18n-name-key="second-custom-field.name"
+    class="com.atlassian.shele.shelePlugin.customfield.SecondCustomField">
+            <description key="second-custom-field.description"/>
+    <resource type="velocity" name="edit" location="/template/second.vm"/>
+        <resource type="velocity" name="view" location="/template/valid.vm"/>
```
Comment (pavel-fando) [inline src/main/resources/atlassian-plugin.xml]: Names of your `.vm` are not describe their purpose. For excample you can create new folder for your custom field and rename files to `edit.vm` and `view.vm`

**Review by pavel-fando — CHANGES_REQUESTED** [review]

File: src/main/java/com/atlassian/shele/shelePlugin/customfield/SecondCustomField.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+import javax.annotation.Nonnull;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static com.atlassian.shele.shelePlugin.utilit.Utilities.CUST_FIELD;
+
+public class SecondCustomField extends GenericTextCFType {
+
+    public SecondCustomField(@ComponentImport CustomFieldValuePersister customFieldValuePersister,
```
Comment (VolodjaGuzh) [inline src/main/java/com/atlassian/shele/shelePlugin/customfield/SecondCustomField.java]: why do you need this constructor here?

File: src/main/java/com/atlassian/shele/shelePlugin/customfield/SecondCustomField.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+        super(customFieldValuePersister, genericConfigManager);
+    }
+
+    @Nonnull
+    @Override
+    public Map<String, Object> getVelocityParameters(final Issue issue, final CustomField field,
+                                                     final FieldLayoutItem fieldLayoutItem) {
+        Map<String, Object> params = new HashMap<>();
+        String reportKey = issue.getReporter().getKey();
+        params.put("reportKey", reportKey);
```
Comment (VolodjaGuzh) [inline src/main/java/com/atlassian/shele/shelePlugin/customfield/SecondCustomField.java]: is it "report key" or "reporter key"?

File: src/main/java/com/atlassian/shele/shelePlugin/customfield/SecondCustomField.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+        return params;
+    }
+
+    @Override
+    public String getValueFromIssue(final CustomField field, final Issue issue) {
+        List<ChangeItemBean> itemBeans = ComponentAccessor.getChangeHistoryManager()
+                .getChangeItemsForField(issue, CUST_FIELD);
```
Comment (VolodjaGuzh) [inline src/main/java/com/atlassian/shele/shelePlugin/customfield/SecondCustomField.java]: does it actually count the number of assignees? what if we perform unassign actions?

File: src/main/java/com/atlassian/shele/shelePlugin/customfield/SecondCustomField.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+import com.atlassian.plugin.spring.scanner.annotation.imports.ComponentImport;
+
+import javax.annotation.Nonnull;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static com.atlassian.shele.shelePlugin.utilit.Utilities.CUST_FIELD;
+
+public class SecondCustomField extends GenericTextCFType {
```
Comment (VolodjaGuzh) [inline src/main/java/com/atlassian/shele/shelePlugin/customfield/SecondCustomField.java]: please add javaDoc for the custom field class and all public methods

File: src/main/java/com/atlassian/shele/shelePlugin/customfield/SimpleCustomField.java [hunk 10]
```diff
@@ -3,23 +3,17 @@
...
 import com.atlassian.jira.issue.customfields.persistence.CustomFieldValuePersister;
-import com.atlassian.jira.security.JiraAuthenticationContext;
 import com.atlassian.plugin.spring.scanner.annotation.imports.ComponentImport;
 
 /**
  * SimpleCustomField module access for change item only
  * amin
  */
-//TODO create customfield with count of assignee into it
 public class SimpleCustomField extends GenericTextCFType {
```
Comment (VolodjaGuzh) [inline src/main/java/com/atlassian/shele/shelePlugin/customfield/SimpleCustomField.java:12]: what is the purpose of this field?

File: src/main/resources/atlassian-plugin.xml [hunk 11]
```diff
@@ -62,7 +62,12 @@
         <resource type="velocity" name="edit" location="/template/edit.vm"/>
         <resource type="velocity" name="view" location="/template/custom.vm"/>
     </customfield-type>
-
+    <customfield-type key="second-custom-field" name="Second custom field" i18n-name-key="second-custom-field.name"
```
Comment (VolodjaGuzh) [inline src/main/resources/atlassian-plugin.xml]: please give some more meaningful name to the field and field class

Comment (VolodjaGuzh) [reply src/main/resources/atlassian-plugin.xml]: the same is for all the directories and views

**Review by VolodjaGuzh — CHANGES_REQUESTED** [review]

File: src/main/java/com/atlassian/shele/shelePlugin/customfield/SecondCustomField.java (hunk 8 above)
Comment (Ivanshelemeh) [reply src/main/java/com/atlassian/shele/shelePlugin/customfield/SecondCustomField.java]: if we remove assignee the count still increments

Comment (VolodjaGuzh) [reply src/main/java/com/atlassian/shele/shelePlugin/customfield/SecondCustomField.java]: so, does it meet the requirements?

Comment (pavel-fando) [reply src/main/java/com/atlassian/shele/shelePlugin/customfield/SecondCustomField.java]: Yes, each change should increase counter by 1

File: src/main/java/com/atlassian/shele/shelePlugin/customfield/CountAssigneeField.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+     *
+     * @param field custom field  object
+     * @param issue current issue on project
+     * @return assignee count
+     */
+    @Override
+    public String getValueFromIssue(final CustomField field, final Issue issue) {
+        List<ChangeItemBean> itemBeans = ComponentAccessor.getChangeHistoryManager()
+                .getChangeItemsForField(issue, CUST_FIELD);
+        int lengthAssignee = itemBeans.size();
```
Comment (VolodjaGuzh) [inline src/main/java/com/atlassian/shele/shelePlugin/customfield/CountAssigneeField.java]: could be simplified to:
`return  String.valueOf(itemBeans.size());`

File: src/main/resources/template/count-assignee-field/view.vm [hunk 13]
```diff
@@ -0,0 +1,3 @@
+<div>
```
Comment (VolodjaGuzh) [inline src/main/resources/template/count-assignee-field/view.vm]: what is the difference between `view.vm` and `view-count-assignee.vm`?

File: src/main/resources/template/edit.vm [hunk 14]
```diff
@@ -7,9 +7,9 @@
     #if($value && ! $value.equals(" "))
```
Comment (VolodjaGuzh) [inline src/main/resources/template/edit.vm:7]: please move this custom field templates to the separate directory as well

File: src/main/java/com/atlassian/shele/shelePlugin/customfield/SimpleCustomField.java [hunk 15]
```diff
@@ -3,23 +3,18 @@
 import com.atlassian.jira.issue.customfields.impl.GenericTextCFType;
 import com.atlassian.jira.issue.customfields.manager.GenericConfigManager;
 import com.atlassian.jira.issue.customfields.persistence.CustomFieldValuePersister;
-import com.atlassian.jira.security.JiraAuthenticationContext;
 import com.atlassian.plugin.spring.scanner.annotation.imports.ComponentImport;
 
 /**
- * SimpleCustomField module access for change item only
- * amin
+ * SimpleCustomField module for examining how create a customfield
```
Comment (VolodjaGuzh) [inline src/main/java/com/atlassian/shele/shelePlugin/customfield/SimpleCustomField.java:9]: is this the correct description or the field should be available for editing only for admins and shown to everybody?
please also rename the field class to reflect its purpose

**Review by VolodjaGuzh — CHANGES_REQUESTED** [review]

**Review by VolodjaGuzh — APPROVED** [review]

**Pull request merged**